packages/core/src/utils/external.ts

```diff
@@ -0,0 +1,58 @@
+import type { RsbuildConfig } from '@rsbuild/core';
+import type { PkgJson } from '../types';
+import type { AutoExternal } from '../types/config';
+import { logger } from './logger';
+
+export const composeAutoExternalConfig = (options: {
+  autoExternal: AutoExternal;
+  pkgJson?: PkgJson;
+  userExternals?: NonNullable<RsbuildConfig['output']>['externals'];
+}): RsbuildConfig => {
+  const { autoExternal, pkgJson, userExternals } = options;
+
+  if (!autoExternal) {
+    return {};
+  }
+
+  if (!pkgJson) {
+    logger.warn(
+      'autoExternal configuration will not be applied due to read package.json failed',
+    );
+    return {};
+  }
+
+  const externalOptions = {
+    dependencies: true,
+    peerDependencies: true,
+    devDependencies: false,
+    ...(autoExternal === true ? {} : autoExternal),
+  };
+
+  // User externals configuration has higher priority than autoExternal
+  // eg: autoExternal: ['react'], user: output: { externals: { react: 'react-1' } }
+  // Only handle the case where the externals type is object, string / string[] does not need to be processed, other types are too complex.
+  const userExternalKeys =
+    userExternals &&
+    Object.prototype.toString.call(userExternals) === '[object Object]'
```

Comment:
Does `userExternals instanceof Object` meets the requirement.

Reply:
```
>  /.a/ instanceof Object => true
>  Object.prototype.toString.call(/.a/) => '[object RegExp]'
```

Reply:
Awesome JaVa$cRipT.